### PR TITLE
Try to fix sql lint true/false negatives

### PR DESCRIPTION
**Pull request by xmo-odoo**
With the current lint we have very similar queries which lead to different flagging:
```python
self.env.cr.execute("SELECT %s FROM table WHERE %s" % (self._table, ''.join(where_list)))
```
triggers a lint failure but
```python
self.env.cr.execute("SELECT %s FROM table WHERE %s" % (self._table, where))
```
does not.

Now there's one part which is a bit unclear: is the first one a false positive or the second one a false negative? It basically boils down to whether what we're allowing is the specific `self._table` being injected directly, or whether `self._table` means all injection is now allowed aka is it ok if *all* injected parameters are `self._...` or *any* of them?

@moylop260 since you were the original author, do you remember the intent? (ignore the existing commits in this PR, basically I assumed the former situation where each parameter was whitelisted but that is quite noisy and unwieldy, which led me to believe your intent had been different).

### CODE REVIEW

**Comment by xmo-odoo** [comment]
@moylop260 also one case is a bit unclear to me:

    # Ignoring for this case: execute("..." % _foo(...))

What's the purpose of this one?

**Comment by moylop260** [comment]
> With the current lint we have very similar queries which lead to different flagging:
```python
self.env.cr.execute("SELECT %s FROM table WHERE %s" % (self._table, ''.join(where_list)))
```
> triggers a lint failure but
```python
self.env.cr.execute("SELECT %s FROM table WHERE %s" % (self._table, where))
```
> does not.


it is an issue I missed consider `astroid.Name`
 - `if isinstance(node, astroid.Name):`

`where` variable is an `astroid.Name`

I only have created the following POC fixing it:
 - https://github.com/odoo/odoo/pull/54012


> Now there's one part which is a bit unclear: is the first one a false positive or the second one a false negative? It basically boils down to whether what we're allowing is the specific self._table being injected directly, or whether self._table means all injection is now allowed aka is it ok if all injected parameters are self._... or any of them?

All private attributes `_*` are silent because of odoo reports with sql views are using this style.
 - `cr.execute("..." % (self._table, self._select(), self._from(), self._group_by(), self._having())`

Check the following examples:
https://github.com/odoo/odoo/blob/8444b476f64d022a51358c8fb374b4107256f27b/addons/account/report/account_invoice_report.py#L182
https://github.com/odoo/odoo/blob/8444b476f64d022a51358c8fb374b4107256f27b/addons/hr_timesheet_attendance/report/hr_timesheet_attendance_report.py#L52
https://github.com/odoo/odoo/blob/8444b476f64d022a51358c8fb374b4107256f27b/addons/membership/report/report_membership.py#L107
https://github.com/odoo/odoo/blob/8444b476f64d022a51358c8fb374b4107256f27b/addons/point_of_sale/report/pos_order_report.py#L107
https://github.com/odoo/odoo/blob/8444b476f64d022a51358c8fb374b4107256f27b/addons/purchase/report/purchase_report.py#L58
https://github.com/odoo/odoo/blob/8444b476f64d022a51358c8fb374b4107256f27b/addons/sale_timesheet/report/project_profitability_report_analysis.py#L190
https://github.com/odoo/odoo/blob/8444b476f64d022a51358c8fb374b4107256f27b/addons/sale/report/report_all_channels_sales.py#L82
https://github.com/odoo/odoo/blob/8444b476f64d022a51358c8fb374b4107256f27b/addons/sale/report/sale_report.py#L139

But if you have
 - `cr.execute("..." % (self._table, variable)`
So, it should raise a sql-injection since that the second param is not a private one.

I have created the following gists for you with sql-injection cases and silent ones and the result of the pylint output using my new PR:
 - https://gist.github.com/moylop260/74b19122d386de0f320816b19dc577d9

**Comment by xmo-odoo** [comment]
> I only have created the following POC fixing it:
> #54012

Seems unnecesary, the first commit of this branch already does that.

> > Now there's one part which is a bit unclear: is the first one a false positive or the second one a false negative? It basically boils down to whether what we're allowing is the specific self._table being injected directly, or whether self._table means all injection is now allowed aka is it ok if all injected parameters are self._... or any of them?
> 
> All private attributes `_*` are silent because of odoo reports with sql views are using this style.
> 
>     * `cr.execute("..." % (self._table, self._select(), self._from(), self._group_by(), self._having())`

Yes I know that.

> But if you have
> 
>     * `cr.execute("..." % (self._table, variable)`
>       So, it should raise a sql-injection since that the second param is not a private one.

At the same time it does make sense, if you are injecting a dynamic table name, chances are high you might need to inject dynamic conditions. Here are the cases the first commit (rejects if any non-whitelisted parameter) finds which the second (reject if *all* non-whitelisted parameters) doesn't:

https://github.com/odoo/odoo/blob/2387699233e2c91d47804a5039925829f9546611/odoo/addons/base/models/ir_sequence.py#L51-L53
https://github.com/odoo/odoo/blob/2387699233e2c91d47804a5039925829f9546611/odoo/addons/base/models/ir_translation.py#L137-L142
https://github.com/odoo/odoo/blob/2387699233e2c91d47804a5039925829f9546611/addons/fleet/report/fleet_report.py#L131

The first one is the only one which isn't *completely* justified and is just a bit lazy.

**Comment by xmo-odoo** [comment]
Also not sure you've seen it, but this is whitelisted:

    # Ignoring for this case: execute("..." % _foo(...))

and I don't understand why, can you explain why specifically underscore-prefixed *function calls*?

**Comment by moylop260** [comment]
Hi @xmo-odoo 

Ok, If I understood well.
You are defining new whitelisted cases.

> - if you are injecting a dynamic table name, chances are high you might need to inject dynamic conditions.

You have a good point
👍 +1

> Also not sure you've seen it, but this is whitelisted:
# Ignoring for this case: execute("..." % _foo(...))

Really all private arguments (starting with `_`) are whitelisted if you are calling a private method, field... there is low chances to use an injected parameter of the user. (I think)
But using your new way this concept is not valid anymore because if there is only one `self._` parameter so the other ones don't matter.

> # execute("xxx" % _(b)) is ok (WHY?)

You have a good point.
The translate method is called `_`
There are high changes to inject from user a translation 😞 
You new concept using just private attributes `self._*` is better than all private `_*` without `self`

👍 +1

**Comment by xmo-odoo** [comment]
> Hi @xmo-odoo
> 
> Ok, If I understood well.
> You are defining new whitelisted cases.

It's a possibility, but after reading your comments it's also true that maybe this is not a good idea because it might let things go through which should not, so I'm a bit conflicted.

All in all after reading your comments and thinking about it, the problem is probably the false *negative* aka the one where there was no warning is wrong, there should probably be a warning in both cases. Still, I'll check with @odony just in case as doing that would require updating a few files.

> > `# execute("xxx" % _(b)) is ok (WHY?)`
> 
> You have a good point.
> The translate method is called `_`
> There are high changes to inject from user a translation disappointed
> You new concept using just private attributes `self._*` is better than all private `_*` without `self`

Yeah I don't think translations would ever be "keywords" so they should probably always be injected as normal parameter. Removing this one is probably a good idea.

**Comment by xmo-odoo** [comment]
@moylop260 after discussing it with Odo, he's fine with "locking down" the linter strictly, and requiring e.g. the use of psycopg2.sql for situations where we need dynamic SQL queries which go beyond just injecting a dynamic table name.

**Comment by pedrobaeza** [comment]
FYI we are using `psycopg2.sql` since some time in openupgradelib and it's better to do it that way and the syntax is straight-forward once you get used to, and allow to compose SQL fragments easily.

See some examples in:

https://github.com/OCA/openupgradelib/pull/205/files
https://github.com/OCA/openupgradelib/pull/201/files
https://github.com/OCA/openupgradelib/pull/199/files
https://github.com/OCA/openupgradelib/pull/198/files
https://github.com/OCA/openupgradelib/pull/197/files
https://github.com/OCA/openupgradelib/pull/195/files

**Comment by xmo-odoo** [comment]
> FYI we are using `psycopg2.sql` since some time in openupgradelib and it's better to do it that way and the syntax is straight-forward once you get used to, and allow to compose SQL fragments easily.
> 
> See some examples in:
> 
> https://github.com/OCA/openupgradelib/pull/205/files
> https://github.com/OCA/openupgradelib/pull/201/files
> https://github.com/OCA/openupgradelib/pull/199/files
> https://github.com/OCA/openupgradelib/pull/198/files
> https://github.com/OCA/openupgradelib/pull/197/files
> https://github.com/OCA/openupgradelib/pull/195/files

We also use it in some places in odoo.

**Comment by xmo-odoo** [comment]
@moylop260 made a few more changes to the linter (e.g. support for "xxx" % <dict> which was not inspected), and found why there was a case for callable prefixed with `_`: reports generally use that pattern to compose their query / view, so allowing those is... convenient. And fixed the cases that tripped up the updated version, either removing unnecessary dsql (rarely) or using `psycopg2.sql` where the dynamic sql composition was warranted.

**Comment by moylop260** [comment]
I got it
My bad I missed to add a comment about.
Thank you for clarification

A PR to enterprise will be required too.

I can help with this if you agree.

**Comment by xmo-odoo** [comment]
> I got it
> My bad I missed to add a comment about.
> Thank you for clarification
> 
> A PR to enterprise will be required too.
> 
> I can help with this if you agree.

No worries, I just wanted to know if you were fine with these changes or if I'd missed things as you're the original author.

The ent changes are simple enough, I'll be fine, but thank you for that as well.

File: odoo/addons/test_lint/tests/_odoo_checker_sql_injection.py [hunk 1]
```diff
@@ -34,41 +34,50 @@ def _get_cursor_name(self, node):
...
+        )
 
     def _check_concatenation(self, node):
-        is_bin_op = False
         if isinstance(node, astroid.BinOp) and node.op in ('%', '+'):
-            # execute("..." % self._table)
-            if (isinstance(node.right, astroid.Attribute) and
-                    not node.right.attrname.startswith('_')):
-                is_bin_op = True
             if isinstance(node.right, astroid.Tuple):
```
Comment (moylop260) [inline odoo/addons/test_lint/tests/_odoo_checker_sql_injection.py:52]: List apply the same `node.right.elts`

```suggestion
            if isinstance(node.right, (astroid.Tuple, astroid.List)):
```

Comment (xmo-odoo) [reply odoo/addons/test_lint/tests/_odoo_checker_sql_injection.py]: Thinking about it more and testing in a python shell, that's not the case, `%` treats a list as a single parameter not as a sequence thereof.

**Comment by xmo-odoo** [comment]
@robodoo r+

File: odoo/addons/test_lint/tests/_odoo_checker_sql_injection.py (hunk 1 above)
Comment (moylop260) [reply odoo/addons/test_lint/tests/_odoo_checker_sql_injection.py:52]: You are right!

**Pull request closed without merging**